src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorSourceDocument.cs

```diff
@@ -2,38 +2,72 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.IO;
 using System.Text;
 
 namespace Microsoft.AspNetCore.Razor.Evolution
 {
     internal class DefaultRazorSourceDocument : RazorSourceDocument
     {
-        private MemoryStream _stream;
+        private readonly string _content;
 
-        public DefaultRazorSourceDocument(MemoryStream stream, Encoding encoding, string filename)
+        public DefaultRazorSourceDocument(string content, Encoding encoding, string filename)
         {
-            if (stream == null)
+            if (content == null)
             {
-                throw new ArgumentNullException(nameof(stream));
+                throw new ArgumentNullException(nameof(content));
             }
 
-            _stream = stream;
+            if (encoding == null)
+            {
+                throw new ArgumentNullException(nameof(encoding));
+            }
+
+            _content = content;
             Encoding = encoding;
             Filename = filename;
         }
 
-        public Encoding Encoding { get; }
+        public override char this[int position] => _content[position];
+
+        public override Encoding Encoding { get; }
 
         public override string Filename { get; }
 
-        public override TextReader CreateReader()
+        public override int Length => _content.Length;
+
+        public override void CopyTo(int sourceIndex, char[] destination, int destinationIndex, int count)
+        {
+            if (!CheckCopyToArguments(sourceIndex, destination, destinationIndex, count))
+            {
+                return;
+            }
+
+            _content.CopyTo(sourceIndex, destination, destinationIndex, count);
+        }
+
+        private bool CheckCopyToArguments(int sourceIndex, char[] destination, int destinationIndex, int count)
         {
-            var copy = new MemoryStream(_stream.ToArray());
+            if (destination == null)
+            {
+                throw new ArgumentNullException(nameof(destination));
+            }
+
+            if (sourceIndex < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(sourceIndex));
+            }
+
+            if (destinationIndex < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(destinationIndex));
+            }
+
+            if (count < 0 || count > Length - sourceIndex || count > destination.Length - destinationIndex)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
 
-            return Encoding == null
-                ? new StreamReader(copy, detectEncodingFromByteOrderMarks: true)
-                : new StreamReader(copy, Encoding);
+            return count > 0;
```

Comment:
Why is this done this way instead of just putting these validations in the enclosing method? We generally want the throwing callstack to be exactly what the user called.